File: Game_Controler/lib/protocol.dart
import 'dart:typed_data';

class MsgTypes {
  static const int move = 0x01;
  static const int buttonDown = 0x02;
  static const int buttonUp = 0x03;
  static const int heartbeat = 0x04;
  static const int batch = 0x05;
}

class Priority {
  static const int high = 2;
  static const int medium = 1;
  static const int low = 0;
}

class Sides {
  static const int left = 0;
  static const int right = 1;
}

enum InputType { move, button }

class InputEvent {
  final InputType type;
  final String side;
  final double x;
  final double y;
  final int index;
  final bool pressed;
  final int priority;
  final int timestamp;

  const InputEvent.move({
    required this.side,
    required this.x,
    required this.y,
    required this.timestamp,
  })  : type = InputType.move,
        index = 0,
        pressed = false,
        priority = Priority.low;

  const InputEvent.button({
    required this.side,
    required this.index,
    required this.pressed,
    required this.timestamp,
  })  : type = InputType.button,
        x = 0,
        y = 0,
        priority = Priority.high;

  int get byteSize => type == InputType.move ? 11 : 7;

  Uint8List toBytes() {
    if (type == InputType.move) {
      return encodeMove(side, x, y, timestamp);
    } else {
      return encodeButton(side, index, pressed, timestamp);
    }
  }
}

Uint8List encodeMove(String side, double x, double y, [int? timestamp]) {
  final data = ByteData(11);
  data.setUint8(0, MsgTypes.move);
  data.setUint8(1, side == 'left' ? Sides.left : Sides.right);
  data.setInt16(2, (x * 32767).round(), Endian.big);
  data.setInt16(4, (y * 32767).round(), Endian.big);
  data.setUint32(
      6, timestamp ?? DateTime.now().millisecondsSinceEpoch, Endian.big);
  return data.buffer.asUint8List();
}

Uint8List encodeButton(String side, int index, bool pressed, [int? timestamp]) {
  final data = ByteData(7);
  data.setUint8(0, pressed ? MsgTypes.buttonDown : MsgTypes.buttonUp);
  data.setUint8(1, side == 'left' ? Sides.left : Sides.right);
  data.setUint8(2, index);
  data.setUint32(
      3, timestamp ?? DateTime.now().millisecondsSinceEpoch, Endian.big);
  return data.buffer.asUint8List();
}

Uint8List encodeHeartbeat([int? timestamp]) {
  final data = ByteData(5);
  data.setUint8(0, MsgTypes.heartbeat);
  data.setUint32(
      1, timestamp ?? DateTime.now().millisecondsSinceEpoch, Endian.big);
  return data.buffer.asUint8List();
}

Uint8List encodeBatch(List<InputEvent> events) {
  if (events.isEmpty) return Uint8List(0);

  int totalSize = 1 + events.fold(0, (sum, e) => sum + e.byteSize);
  final buffer = ByteData(totalSize);

  buffer.setUint8(0, MsgTypes.batch);
  buffer.setUint8(1, events.length);

  int offset = 2;
  for (var event in events) {
    final bytes = event.toBytes();
    for (int i = 0; i < bytes.length; i++) {
      buffer.setUint8(offset + i, bytes[i]);
    }
    offset += bytes.length;
  }

  return buffer.buffer.asUint8List();
}

File: Backend_controler/protocol.js
const MSG_TYPES = {
  MOVE: 0x01,
  BUTTON_DOWN: 0x02,
  BUTTON_UP: 0x03,
  HEARTBEAT: 0x04,
  BATCH: 0x05,
};

const SIDES = {
  left: 0,
  right: 1,
};

const SIDE_NAMES = ["left", "right"];

function decodeMessage(buffer) {
  if (!buffer || buffer.length < 5) return null;

  try {
    const view = buffer instanceof Buffer ? buffer : Buffer.from(buffer);
    const type = view[0];

    switch (type) {
      case MSG_TYPES.MOVE: {
        if (view.length < 11) return null;
        const side = SIDE_NAMES[view[1]] || "left";
        const x = view.readInt16BE(2) / 32767;
        const y = view.readInt16BE(4) / 32767;
        const timestamp = view.readUInt32BE(6);
        return { type: "move", side, x, y, timestamp };
      }
      case MSG_TYPES.BUTTON_DOWN: {
        if (view.length < 7) return null;
        const side = SIDE_NAMES[view[1]] || "left";
        const index = view[2];
        const timestamp = view.readUInt32BE(3);
        return { type: "button_down", side, index, timestamp };
      }
      case MSG_TYPES.BUTTON_UP: {
        if (view.length < 7) return null;
        const side = SIDE_NAMES[view[1]] || "left";
        const index = view[2];
        const timestamp = view.readUInt32BE(3);
        return { type: "button_up", side, index, timestamp };
      }
      case MSG_TYPES.HEARTBEAT: {
        if (view.length < 5) return null;
        const timestamp = view.readUInt32BE(1);
        return { type: "heartbeat", timestamp };
      }
      case MSG_TYPES.BATCH: {
        if (view.length < 3) return null;
        const count = view[1];
        const events = [];
        let offset = 2;

        for (let i = 0; i < count && offset < view.length; i++) {
          const msgType = view[offset];
          let event = null;

          switch (msgType) {
            case MSG_TYPES.MOVE: {
              if (offset + 10 < view.length) {
                const side = SIDE_NAMES[view[offset + 1]] || "left";
                const x = view.readInt16BE(offset + 2) / 32767;
                const y = view.readInt16BE(offset + 4) / 32767;
                const timestamp = view.readUInt32BE(offset + 6);
                event = { type: "move", side, x, y, timestamp };
                offset += 11;
              }
              break;
            }
            case MSG_TYPES.BUTTON_DOWN: {
              if (offset + 6 < view.length) {
                const side = SIDE_NAMES[view[offset + 1]] || "left";
                const index = view[offset + 2];
                const timestamp = view.readUInt32BE(offset + 3);
                event = { type: "button_down", side, index, timestamp };
                offset += 7;
              }
              break;
            }
            case MSG_TYPES.BUTTON_UP: {
              if (offset + 6 < view.length) {
                const side = SIDE_NAMES[view[offset + 1]] || "left";
                const index = view[offset + 2];
                const timestamp = view.readUInt32BE(offset + 3);
                event = { type: "button_up", side, index, timestamp };
                offset += 7;
              }
              break;
            }
            default:
              offset = view.length;
          }

          if (event) events.push(event);
        }

        return { type: "batch", events };
      }
      default:
        return null;
    }
  } catch (e) {
    return null;
  }
}

module.exports = {
  MSG_TYPES,
  SIDES,
  decodeMessage,
};
